### PR TITLE
Add ability to pass referrer when opening new window.

### PR DESCRIPTION
**Pull request by auniverseaway**
Please ensure your pull request adheres to the following guidelines:
- [X] make sure to link the related issues in this description
- [ ] when merging / squashing, make sure the fixed issue references are visible in the commits, for easy compilation of release notes

## Related Issues
GH-136

Thanks for contributing!

### CODE REVIEW

File: src/extension/module.js [hunk 1]
```diff
@@ -1246,9 +1247,10 @@
               text: (titleI18n && titleI18n[language]) || title,
               action: () => {
                 if (url) {
-                  const target = url.startsWith('/')
+                  const baseTarget = url.startsWith('/')
                     ? new URL(url, `https://${innerHost}/`)
                     : url;
+                  const target = passReferrer ? `${baseTarget}?Referrer=${encodeURIComponent(location.href)}` : baseTarget;
```
Comment (tripodsan) [inline src/extension/module.js:1253]: add the referrer via `URL.searchParams` otherwise you might have two `?` in the string.

Comment (rofe) [inline src/extension/module.js:1253]: I'd prefer a lowercase parameter name (like in the issue description): `referrer`

**Review by rofe — CHANGES_REQUESTED** [review]
Thanks for the contribution @auniverseaway! Please don't forget to also extend the [`config.schema.json`](https://github.com/adobe/helix-sidekick-extension/blob/mv3/docs/config.schema.json).

**Comment by rofe** [comment]
JSDoc and JSON schema will also need to be updated.

**Comment by auniverseaway** [comment]
I have some E2E and tests to write tomorrow. It seems a bit suspect that URL would write out to a string. The original solution had that for relative links... I'm using it for both. Will take a look in the morning.

File: src/extension/module.js [hunk 2]
```diff
@@ -1247,10 +1247,11 @@
...
               action: () => {
                 if (url) {
-                  const baseTarget = url.startsWith('/')
-                    ? new URL(url, `https://${innerHost}/`)
-                    : url;
-                  const target = passReferrer ? `${baseTarget}?Referrer=${encodeURIComponent(location.href)}` : baseTarget;
+                  const target = url.startsWith('/') ? new URL(url, `https://${innerHost}/`) : new URL(url);
+                  if (passReferrer) {
+                    const { searchParams } = target;
+                    searchParams.append('referrer', encodeURIComponent(location.href));
```
Comment (tripodsan) [inline src/extension/module.js:1253]: you don't need to encode the argument. the `searchParams` do this automatically.

**Review by tripodsan — CHANGES_REQUESTED** [review]

File: src/extension/module.js (hunk 1 above)
Comment (auniverseaway) [reply src/extension/module.js]: Should be fixed.

Comment (auniverseaway) [reply src/extension/module.js]: Should be added.

File: src/extension/module.js (hunk 2 above)
Comment (auniverseaway) [reply src/extension/module.js]: Should be fixed. Verified with a branch on milo.

**Comment by auniverseaway** [comment]
@rofe @tripodsan Everything should be resolved.

Regarding tests: I saw (to my knowledge) there wasn't a test for a custom URL. This seems to be where WTR / Mocha / unit tests falls down... we cannot validate a click + new window + query string. Solenium / Playwright would be great here.

If either of you have ideas on how I can add a test using the existing stack, I'd love to get something written. Especially seeing that module.js is sitting at ~67% coverage at the moment.

I **_did_** validate that everything works as expected on Milo.

**Review by tripodsan — APPROVED** [review]

**Comment by rofe** [comment]
> @rofe @tripodsan Everything should be resolved.
> 
> Regarding tests: I saw (to my knowledge) there wasn't a test for a custom URL. This seems to be where WTR / Mocha / unit tests falls down... we cannot validate a click + new window + query string. Solenium / Playwright would be great here.
> 
> If either of you have ideas on how I can add a test using the existing stack, I'd love to get something written. Especially seeing that module.js is sitting at ~67% coverage at the moment.
> 
> I **_did_** validate that everything works as expected on Milo.

@auniverseaway I recently added (puppeteer-based) UI tests for plugins using `url` and `event`. We could add one for the `passReferrer` case, too. See https://github.com/adobe/helix-sidekick-extension/blob/mv3/test/module.test.js#L191 and following.

**Pull request merged**